### ISSUE TITLE
[enhancement] Better layout algorithm

### ISSUE THREAD

**Issue by dvkndn** [TablePlus/diagram-plugin#15]
> There was this another issue in the previous version _(which I have previously installed)_ where the line was overlapped by table.
> 
> Maybe it might have been fixed in the newer version.
> 
> ![CleanShot 2020-10-20 at 06 37 20@2x](https://user-images.githubusercontent.com/29014463/96873941-2fd26e00-1493-11eb-9c20-c05d960c6774.png)
> 
> _Originally posted by @mittalyashu in https://github.com/TablePlus/diagram-plugin/issues/5#issuecomment-714471708_

> > At the moment we choose to place the Tables over the References.

> As far I understand the canvas _(well, I don't even know how this plugin is build)_ has infinite space so just push the table bit further away so that the references doesn't over lap the tables.
> 
> And I mean by the above image.
> 
> NOTE: The above image was taken after the table arrangement was done by plugin automatically, I didn't change their positions.
> 
> <img width="119" alt="CleanShot 2020-10-23 at 09 18 49@2x" src="https://user-images.githubusercontent.com/29014463/96954315-c6914000-1510-11eb-9787-f537368430d4.png">
> 
> Not this 👆
> 
> _Originally posted by @mittalyashu in https://github.com/TablePlus/diagram-plugin/issues/5#issuecomment-714892046_
> 
> As far I understand the canvas (well, I don't even know how this plugin is build) has infinite space so just push the table bit further away so that the references doesn't over lap the tables.

> This is a very reasonable approach :D Unfortunately our current "layout" algorithm is not quite as good. Let us improve it. You are right that the layout should put the tables to minimize the overlaps
> 
> _Originally posted by @dvkndn in https://github.com/TablePlus/diagram-plugin/issues/5#issuecomment-714893168_

**Comment by mittalyashu** [comment]
@dvkndn That's not really an infinite canvas, it only expands to X+ and Y+ axis

![CleanShot 2020-12-04 at 09 54 23](https://user-images.githubusercontent.com/29014463/101121715-d1aeb400-3616-11eb-9ddb-1525a531387d.gif)

**Comment by dvkndn** [comment]
Yes :( The fact that it is not an infinite canvas is causing many issues now. Basically we need to make it infinite first (or just that it's really really buig)

**Comment by migueltorrescosta** [comment]
Another layout issue: When plotting huge databases we might get connected tables on opposite ends of the layout. Is there a way to find better layouts?
I used different layout algorithms when facing the same issue using NetworkX, I would be happy to do a MR for this if you can give me some guidance as to where / how the current calculations are done

**Comment by thien-do** [comment]
oh great @migueltorrescosta !

>  Is there a way to find better layouts?

We are looking into this. But the biggest issue is that we need to support the infinite canvas first. It would make all of these things much easier. To be specific, the (0, 0) should be the center, not at top left like this

> I would be happy to do a MR for this if you can give me some guidance as to where / how the current calculations are done

Unfortunately at the moment the original source code is not completely open :(

**Comment by migueltorrescosta** [comment]
I understand. If there's anything I could do to speed up the process do ask, it would make the generator that much more valuable :)

**Comment by s2k** [comment]
I'd also appreciate an improved layout algorithm.
I wonder if it's possible to reduce the number of line crossings (as much as possible).
For example GraphViz (https://www.graphviz.org) does it, other libraries can surely do it as well. (I only mention this one, because I'm comfortable with it.)

This super simple example shows what I mean:

<img width="500" alt="table-diagram" src="https://user-images.githubusercontent.com/118248/160647630-bc74aaab-f3b2-4752-b1d9-6a4e4909f8c0.png">

The connections `line_items ➙ orders` and `line_items ➙ carts` cross each other, even though it's (topologically) not necessary.
